sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,315 @@
+# -*- coding: utf-8 -*-
+"""K-medoids clustering"""
+
+# Authors: Timo Erkkilä <timo.erkkila@gmail.com>
+#          Antti Lehmussola <antti.lehmussola@gmail.com>
+#          Kornel Kiełczewski <kornel.k@plusnet.pl>
+# License: BSD 3 clause
+
+import warnings
+
+import numpy as np
+
+from ..base import BaseEstimator, ClusterMixin, TransformerMixin
+from ..metrics.pairwise import PAIRWISE_DISTANCE_FUNCTIONS
+from ..utils import check_array, check_random_state
+from ..utils.validation import check_is_fitted
+
+
+class KMedoids(BaseEstimator, ClusterMixin, TransformerMixin):
+    """
+    k-medoids class.
+
+    Parameters
+    ----------
+    n_clusters : int, optional, default: 8
+        How many medoids. Must be positive.
+
+    distance_metric : string, optional, default: 'euclidean'
+        What distance metric to use.
+
+    clustering : {'pam'}, optional, default: 'pam'
+        What clustering mode to use.
+
+    init : {'random', 'heuristic'}, optional, default: 'heuristic'
+        Specify medoid initialization.
+
+    max_iter : int, optional, default : 300
+        Specify the maximum number of iterations when fitting.
+
+    random_state : int, optional, default: None
+        Specify random state for the random number generator.
```

Comment:
The doc example should preferably demonstrate how the different distance metrics can be used